hack/make/.binary

```diff
@@ -16,61 +16,88 @@ source "${MAKEDIR}/.go-autogen"
-				windows/amd64)
-					export CC="${CC:-x86_64-w64-mingw32-gcc}"
-					export CGO_ENABLED=1
-					;;
-				linux/arm)
-					case "${GOARM}" in
-						5)
-							export CC="${CC:-arm-linux-gnueabi-gcc}"
-							export CGO_ENABLED=1
-							export CGO_CFLAGS="-march=armv5t"
-							export CGO_CXXFLAGS="-march=armv5t"
-							;;
-						6)
-							export CC="${CC:-arm-linux-gnueabi-gcc}"
-							export CGO_ENABLED=1
-							export CGO_CFLAGS="-march=armv6"
-							export CGO_CXXFLAGS="-march=armv6"
-							;;
-						7)
-							export CC="${CC:-arm-linux-gnueabihf-gcc}"
-							export CGO_ENABLED=1
-							export CGO_CFLAGS="-march=armv7-a"
-							export CGO_CXXFLAGS="-march=armv7-a"
-							;;
-						*)
-							export CC="${CC:-arm-linux-gnueabihf-gcc}"
-							export CGO_ENABLED=1
-							;;
-					esac
-					;;
-				linux/arm64)
-					export CC="${CC:-aarch64-linux-gnu-gcc}"
-					export CGO_ENABLED=1
-					;;
-				linux/amd64)
-					export CC="${CC:-x86_64-linux-gnu-gcc}"
-					export CGO_ENABLED=1
-					;;
-				linux/ppc64le)
-					export CC="${CC:-powerpc64le-linux-gnu-gcc}"
-					export CGO_ENABLED=1
-					;;
-				linux/s390x)
-					export CC="${CC:-s390x-linux-gnu-gcc}"
-					export CGO_ENABLED=1
-					;;
-			esac
+	# only necessary for non-sandboxed invocation where TARGETPLATFORM is empty
+	if [ -z "$TARGETPLATFORM" ]; then
+		TARGETPLATFORM="$(go env GOOS)/$(go env GOARCH)"
+		if [ -n "$(go env GOARM)" ]; then
+			TARGETPLATFORM+="/v$(go env GOARM)"
+		elif [ -n "$(go env GOAMD64)" ] && [ "$(go env GOAMD64)" != "v1" ]; then
+			TARGETPLATFORM+="/$(go env GOAMD64)"
 		fi
 	fi
 
+	iscross=0
+	sandboxed=0
+	if command -v xx-info > /dev/null 2>&1; then
+		sandboxed=1
+		iscross=$(xx-info is-cross && echo 1 || echo 0)
+	elif [ "$(go env GOOS)/$(go env GOARCH)" != "$(go env GOHOSTOS)/$(go env GOHOSTARCH)" ]; then
+		iscross=1
+	fi
+
+	if [[ "$sandboxed" = "0" ]] && [[ "$iscross" = "1" ]]; then
+		case "$(go env GOOS)/$(go env GOARCH)" in
+			windows/amd64)
+				export CC="${CC:-x86_64-w64-mingw32-gcc}"
+				export CGO_ENABLED=1
+				;;
+			linux/arm)
+				case "$(go env GOARM)" in
+					5)
+						export CC="${CC:-arm-linux-gnueabi-gcc}"
+						export CGO_ENABLED=1
+						;;
+					6)
+						export CC="${CC:-arm-linux-gnueabi-gcc}"
+						export CGO_ENABLED=1
+						;;
+					7)
+						export CC="${CC:-arm-linux-gnueabihf-gcc}"
+						export CGO_ENABLED=1
+						;;
+					*)
+						export CC="${CC:-arm-linux-gnueabihf-gcc}"
+						export CGO_ENABLED=1
+						;;
+				esac
+				;;
+			linux/arm64)
+				export CC="${CC:-aarch64-linux-gnu-gcc}"
+				export CGO_ENABLED=1
+				;;
+			linux/amd64)
+				export CC="${CC:-x86_64-linux-gnu-gcc}"
+				export CGO_ENABLED=1
+				;;
+			linux/ppc64le)
+				export CC="${CC:-powerpc64le-linux-gnu-gcc}"
+				export CGO_ENABLED=1
+				;;
+			linux/s390x)
+				export CC="${CC:-s390x-linux-gnu-gcc}"
+				export CGO_ENABLED=1
+				;;
+		esac
+	fi
+
+	if [ "$iscross" = "1" ]; then
```

Comment:
Can we make this comment a little more useful? Something like "since 32-bit ARM shares a toolchain, we must set the `-march` flag to select the correct subarchitecture"